packages/alfa-rules/src/helpers/predicate-builder.ts

```diff
@@ -0,0 +1,160 @@
+import { getRole, Role } from "@siteimprove/alfa-aria";
+import { BrowserSpecific } from "@siteimprove/alfa-compatibility";
+import { Device } from "@siteimprove/alfa-device";
+import {
+  Element,
+  getElementNamespace,
+  getInputType,
+  InputType,
+  isElement as isElt,
+  Namespace,
+  Node
+} from "@siteimprove/alfa-dom";
+import { Predicate } from "@siteimprove/alfa-util";
+
+const {
+  BinOp: { and, or }
+} = BrowserSpecific;
+
+class PredicateBuilder<T, U extends T = T> {
+  public readonly predicate: Predicate<T, U>;
+
+  public constructor(predicate: Predicate<T, U> = () => true) {
+    this.predicate = predicate;
+  }
+
+  public and<V extends U = U>(
+    predicate: Predicate<U, V>
+  ): PredicateBuilder<T, V> {
+    return new PredicateBuilder(v => this.predicate(v) && predicate(v));
+  }
+
+  public browserSpecific(): BrowserSpecificPredicateBuilder<T, U> {
+    return new BrowserSpecificPredicateBuilder(this.predicate);
+  }

```

Comment:
This is ugly 😞 
`and` should be overloaded with `predicate: BrowserSpecificPredicate<U, V>` and handle this. But it does not look like I can test the type between `Predicate` and `BSPredicate` to choose what to do…